File: client/test/unit/product_service_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:client/data/models/product/create_product_request.dart';

void main() {
  group('Tests Esenciales de Product', () {

    group('Modelo de Producto', () {
      test('debería crear CreateProductRequest correctamente', () {
        // Act
        final request = CreateProductRequest(
          name: 'Producto Test',
          price: 25000,
          stock: 100,
          categoryId: 'cat1',
          providerId: 'prov1',
        );

        // Assert
        expect(request.name, 'Producto Test');
        expect(request.price, 25000.0);
        expect(request.stock, 100);
        expect(request.categoryId, 'cat1');
        expect(request.providerId, 'prov1');
      });

      test('debería validar datos del producto', () {
        // Act
        final request = CreateProductRequest(
          name: 'Test Product',
          price: 100.0,
          stock: 10,
          categoryId: 'cat1',
          providerId: 'prov1',
        );
        
        // Assert
        expect(request.price > 0, isTrue);
        expect(request.name.isNotEmpty, isTrue);
      });
    });
  });
}

File: server/tests/unit/services/productService.test.ts
import { describe, it, expect, beforeAll, afterAll, beforeEach } from 'vitest';
import { setupTestDB, teardownTestDB, clearDatabase } from '../../setup.js';
import productService from '../../../src/services/productService.js';
import Product from '../../../src/models/productModel.js';
import { IProduct } from '../../../src/interfaces/models.interface.js';
import { Types } from 'mongoose';

describe('Tests Unitarios de ProductService', () => {
    let testUserId: string;
    let testCategoryId: string;
    let testProviderId: string;

    beforeAll(async () => {
        await setupTestDB();
    });

    afterAll(async () => {
        await teardownTestDB();
    });

    beforeEach(async () => {
        await clearDatabase();
        
        // Crear IDs válidos para los tests
        testUserId = new Types.ObjectId().toString();
        testCategoryId = new Types.ObjectId().toString();
        testProviderId = new Types.ObjectId().toString();
    });

    describe('Creación de Producto', () => {
        it('debería fallar con usuario inexistente', async () => {
            // Arrange
            const productData: IProduct = {
                name: 'Producto Test',
                price: 100,
                stock: 50,
                categoryId: new Types.ObjectId(testCategoryId),
                providerId: new Types.ObjectId(testProviderId),
                userId: new Types.ObjectId(testUserId),
                imageUrl: null,
                createdAt: new Date(),
                deletedAt: null
            };

            // Act
            const result = await productService.createProduct(productData, testUserId);

            // Assert
            expect(result.success).toBe(false);
            expect(result.message).toBeDefined();
        });
    });

    describe('Obtener Productos', () => {
        it('debería fallar al obtener productos con usuario inexistente', async () => {
            // Act
            const result = await productService.getProducts(testUserId);

            // Assert
            expect(result.success).toBe(false);
            expect(result.error).toBe('not_found');
            expect(result.message).toBe('User not found');
        });
    });
});